# Reading and using BWA

##Preprocessing

The command !apt-get install -y bwa installs the BWA (Burrows-Wheeler Aligner) software on your system, enabling you to perform sequence alignment of reads against a reference genome.

In [1]:
# Install BWA
!apt-get install -y bwa

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
bwa is already the newest version (0.7.17-6).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.


Unzips the Reference file

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
# Unzip the uploaded file
!unzip -o "/content/drive/MyDrive/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip" -d "/content/"

Archive:  /content/drive/MyDrive/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta  
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta  
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_contigs.agp  


In [ ]:
# Unzip the uploaded file
# !unzip "/content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip" -d "/content/"  #Gauranga drive
!unzip "/content/drive/MyDrive/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip" -d "/content/"  #Siddhesh  drive

Archive:  /content/drive/MyDrive/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip
replace /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [6]:
# List the files in the current directory
!ls -lh

total 8.6M
drwx------ 5 root root 4.0K Sep 23 10:35 drive
-rw-r--r-- 1 root root   53 Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.agp
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.fasta
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta
drwxr-xr-x 1 root root 4.0K Sep 19 13:25 sample_data


The output of the `!ls -lh` command lists the files and directories in the current working directory with detailed information. Here's a breakdown of each line:

 General Information

-   **Total 12MB**: The total size of all files in the current directory is 12 Megabytes.

 Summary of Differences

| **Feature** | **Contigs (`contigs.fasta`)** | **Supercontigs (`supercontigs.fasta`)** |
| --- | --- | --- |
| **Definition** | Continuous sequences assembled from reads | Linked contigs with known order/orientation |
| **Length** | Shorter | Longer |
| **Gaps** | No gaps | May contain gaps (`N` bases) |
| **Assembly Level** | Initial stage of assembly | Higher level, linked contigs (scaffolds) |
| **Use Cases** | Local sequence analysis | Genome structure and gene prediction |
| **Completeness** | Less complete representation of genome | More complete, near full-chromosome length |

Installing both `bwa` and `samtools` together allows you to perform a complete workflow for analyzing sequencing data:

1.  **Align Reads**: Use `bwa` to align your sequencing reads to a reference genome.
2.  **Process Alignments**:
    -   Convert the resulting SAM file to BAM format using `samtools`.
    -   Sort and index the BAM file for efficient access and analysis.
    -   Use `samtools` to generate summary statistics, view alignments, and prepare data for variant calling or visualization.

-   **Indexing**: Preprocessing the reference genome to create data structures (BWT, SA) that enable fast searching and alignment.
-   **Example**: A read is aligned to the reference using the index, which helps BWA quickly find the correct position without scanning the entire genome.
-   **Efficiency**: Indexing significantly speeds up the alignment process by reducing the number of comparisons and memory usage.

In [5]:
# Install BWA and samtools
!apt-get install -y bwa samtools

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
bwa is already the newest version (0.7.17-6).
The following additional packages will be installed:
  libhts3 libhtscodecs2
Suggested packages:
  cwltool
The following NEW packages will be installed:
  libhts3 libhtscodecs2 samtools
0 upgraded, 3 newly installed, 0 to remove and 49 not upgraded.
Need to get 963 kB of archives.
After this operation, 2,270 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]
Fetched 963 kB in 1s (1,038 kB/s)
Selecting previously unselected package libhtscodecs2:amd64.
(Reading database ... 123616 files and directories currently installed.)
Preparing to unpack .../libhtscodecs2_1.1.1-3_amd

In [10]:
# index the reference
!bwa index /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta

[bwa_index] Pack FASTA... 0.04 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 1.67 seconds elapse.
[bwa_index] Update BWT... 0.03 sec
[bwa_index] Pack forward-only FASTA... 0.03 sec
[bwa_index] Construct SA from BWT and Occ... 0.72 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta
[main] Real time: 2.553 sec; CPU: 2.479 sec


Here's a brief description of each file:

1.  **`.fasta.amb`**: This file contains information about ambiguous bases in the reference genome. It helps the aligner handle ambiguous nucleotide sequences (like N's) efficiently.

2.  **`.fasta.ann`**: This file is an annotation file that contains information about the sequences, including their names and lengths.

3.  **`.fasta.bwt`**: This is the Burrows-Wheeler Transform index file. It is used to efficiently align reads to the reference genome. The BWT is a reversible transformation that enables efficient data compression and search.

4.  **`.fasta.pac`**: This file contains the packed version of the reference sequence, stored in a compressed binary format to reduce the file size.

5.  **`.fasta.sa`**: This is the suffix array index file. It is used in conjunction with the `.bwt` file to efficiently locate substrings in the reference sequence during alignment.

The following code aligns sequencing reads from the FASTQ file to the reference genome using the `bwa mem` algorithm and outputs the alignments in a SAM file.

What Happens During the Execution ?

1.  **Alignment Process**:

    -   The `bwa mem` algorithm reads the input FASTQ file (`SRR30683486.fastq.gz`) and aligns each read to the reference genome (`mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta`).
    -   It uses the index files created from the reference genome (generated using the `bwa index` command) to find the best alignment positions quickly and efficiently.
2.  **Output in SAM Format**:

    -   The aligned reads, along with detailed alignment information (position, quality, mismatches, etc.), are saved in the `aligned_reads.sam` file.
    -   This SAM file is a standard format used to store alignment information, and it can be converted to BAM format, sorted, and indexed for further analysis.

In [11]:
# Align the reads with the reference genome using bwa mem
# Gauranga
# !bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta \
# /content/drive/MyDrive/Academic/CMI\ DS\ 23-25/IP/Meeting\ 4:\ 23\ 09\ 2024/SRR30683486.fastq.gz \
# > /content/aligned_reads.sam

!bwa mem /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta \
/content/drive/MyDrive/IP/Meeting\ 4:\ 23\ 09\ 2024/SRR30679840.fastq.gz \
> /content/aligned_reads.sam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 69446 sequences (10000289 bp)...
[M::process] read 70744 sequences (10000130 bp)...
[M::mem_process_seqs] Processed 69446 reads in 5.823 CPU sec, 5.767 real sec
[M::process] read 71080 sequences (10000222 bp)...
[M::mem_process_seqs] Processed 70744 reads in 4.969 CPU sec, 5.042 real sec
[M::process] read 70576 sequences (10000163 bp)...
[M::mem_process_seqs] Processed 71080 reads in 5.062 CPU sec, 5.140 real sec
[M::process] read 71326 sequences (10000297 bp)...
[M::mem_process_seqs] Processed 70576 reads in 5.927 CPU sec, 6.414 real sec
[M::process] read 71296 sequences (10000276 bp)...
[M::mem_process_seqs] Processed 71326 reads in 4.671 CPU sec, 4.654 real sec
[M::process] read 70520 sequences (10000205 bp)...
[M::mem_process_seqs] Processed 71296 reads in 4.803 CPU sec, 4.483 real sec
[M::process] read 69880 sequences (10000259 bp)...
[M::mem_process_seqs] Processed 70520 reads in 6.140 CPU sec, 5.960 real sec
[M::pr

## Reading the results of BWA

### Converting to BAM

In [13]:
# Install samtools
!apt-get install -y samtools

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
samtools is already the newest version (1.13-4).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.


In [14]:
# Convert SAM to BAM formata
!samtools view -Sb /content/aligned_reads.sam > /content/aligned_reads.bam

In [20]:
# Sort the BAM file
!samtools sort /content/aligned_reads.bam -o /content/sorted_reads.bam

Check the basic statistics of the BAM file to get an overview of the alignment.

In [17]:
! samtools flagstat /content/aligned_reads.bam

1474270 + 0 in total (QC-passed reads + QC-failed reads)
1466814 + 0 primary
0 + 0 secondary
7456 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
1462802 + 0 mapped (99.22% : N/A)
1455346 + 0 primary mapped (99.22% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


Alignment Quality

In [18]:
! samtools view /content/aligned_reads.bam | awk '{if($5>=30) count++} END {print "Reads with MAPQ >= 30:", count}'

Reads with MAPQ >= 30: 1437348


In [21]:
! samtools quickcheck -v /content/aligned_reads.bam

In [23]:
# Sort the BAM file
! samtools sort /content/aligned_reads.bam -o /content/sorted_reads.bam

In [24]:
! samtools index /content/sorted_reads.bam

Mark Duplicates Using Picard

In [25]:
! samtools markdup /content/sorted_reads.bam /content/dedup_reads.bam

In [26]:
! samtools flagstat /content/dedup_reads.bam

1474270 + 0 in total (QC-passed reads + QC-failed reads)
1466814 + 0 primary
0 + 0 secondary
7456 + 0 supplementary
368948 + 0 duplicates
368948 + 0 primary duplicates
1462802 + 0 mapped (99.22% : N/A)
1455346 + 0 primary mapped (99.22% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


In [29]:
! samtools view -f 1024 /content/dedup_reads.bam | head


SRR30679840.294142	1024	Supercontig_2.1	1	60	4S146M	*	0	0	GTCGTTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACTCGGTCGTCTCCGAACTTAACGGCGACCCTAAGGTTGACGACGGACCCAGCAGTGATGCTAATCTCAGCGCTCCGCTGACCCCTCAGCAAAGGGCTTGGCTCAA	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:1	MD:Z:42G103	AS:i:141	XS:i:0
SRR30679840.12160	1040	Supercontig_2.1	1	60	18S132M	*	0	0	AACCAGGGAGATACGTCGTTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGTCGTCTCCGAACTTAACGGCGACCCTAAGGTTGACGACGGACCCAGCAGTGATGCTAATCTCAGCGCTCCGCTGACCCCTCAGCAA	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:0	MD:Z:132	AS:i:132	XS:i:0
SRR30679840.358396	1040	Supercontig_2.1	1	60	18S132M	*	0	0	AACCAGGGAGATACGTCGTTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGTCGTCTCCGAACTTAACGGCGACCCTAAGGTTGACGACGGACCCAGCAGTGATGCTAATCTCAGCGCTCCGCTGACCCCTCAGCAA	???

In [30]:
# List the output files
!ls -lh /content/

total 700M
-rw-r--r-- 1 root root  84M Sep 23 10:53 aligned_reads.bam
-rw-r--r-- 1 root root 531M Sep 23 10:46 aligned_reads.sam
-rw-r--r-- 1 root root  35M Sep 23 11:00 dedup_reads.bam
drwx------ 5 root root 4.0K Sep 23 10:35 drive
-rw-r--r-- 1 root root   53 Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.agp
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.fasta
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta
-rw-r--r-- 1 root root   12 Sep 23 10:43 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta.amb
-rw-r--r-- 1 root root   79 Sep 23 10:43 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta.ann
-rw-r--r-- 1 root root 4.3M Sep 23 10:43 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta.bwt
-rw-r--r-- 1 root root 1.1M Sep 23 10:43 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta.pac
-rw-r--r-- 1 root root 2.2M Sep 23 10:43 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta.sa


In [28]:
!pip install pysam

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.0/22.0 MB 35.5 MB/s eta 0:00:00


In [31]:
import pysam

# Open the BAM file
bamfile = pysam.AlignmentFile("/content/sorted_reads.bam", "rb")

# Counter to limit the number of reads printed
counter = 0

# Iterate over reads in the BAM file
for read in bamfile.fetch():
    # Print read information
    print(f"Read Name: {read.query_name}")
    print(f"Reference Name: {read.reference_name}")
    print(f"Alignment Start: {read.reference_start}")
    print(f"Mapping Quality: {read.mapping_quality}")
    print(f"Query Sequence: {read.query_sequence}")
    print(f"Query Quality: {read.query_qualities}")
    print("-" * 50)

    # Increment the counter
    counter += 1

    # Stop after printing 5 reads
    if counter == 5:
        break

bamfile.close()

Read Name: SRR30679840.1834
Reference Name: Supercontig_2.1
Alignment Start: 0
Mapping Quality: 60
Query Sequence: TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGT
Query Quality: array('B', [30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30])
--------------------------------------------------
Read Name: SRR30679840.12160
Reference Name: Supercontig_2.1
Alignment Start: 0
Mapping Quality: 60
Query Sequence: TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGTCGTCTCCGAACTT
Query Quality: array('B', [30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30])
--------------------------------------------------
Read Name: SRR30679840.16549
Reference Name: Supercontig_2.1
Alignment Sta

In [ ]:
# Install samtools and bcftools
!apt-get install -y samtools bcftools

In [ ]:
# Index the reference genome
!samtools faidx /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta

In [ ]:
# Generate a pileup file from the BAM file
!samtools mpileup -f /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta /content/sorted_reads.bam > /content/reads.pileup

In [ ]:
# Call variants using bcftools
!bcftools call -c -v -Ov -o /content/variants.vcf /content/reads.pileup

In [ ]:
import pysam

# Open the BAM file
bamfile = pysam.AlignmentFile("/content/sorted_reads.bam", "rb")

# Iterate over reads in the BAM file
for read in bamfile.fetch():
    # Print basic alignment information
    print(f"Read Name: {read.query_name}")
    print(f"Reference Name: {read.reference_name}")
    print(f"Alignment Start: {read.reference_start}")
    print(f"Mapping Quality: {read.mapping_quality}")
    print(f"CIGAR: {read.cigarstring}")
    print(f"NM Tag (Edit Distance): {read.get_tag('NM')}")

    # Check for MD tag and print mismatches
    if read.has_tag("MD"):
        md_tag = read.get_tag("MD")
        print(f"MD Tag (Mismatches): {md_tag}")

    # Display the query sequence and aligned positions
    print(f"Query Sequence: {read.query_sequence}")
    print(f"Aligned Pairs (query, reference):")
    for qpos, rpos in read.get_aligned_pairs():
        # Check if there is a mismatch
        if rpos is not None and qpos is not None and read.query_sequence[qpos] != read.get_reference_sequence()[rpos]:
            print(f"  Read base: {read.query_sequence[qpos]}, Reference base: {read.get_reference_sequence()[rpos]}, Position: {rpos}")

    print("-" * 50)

    # Limit output to the first read for demonstration purposes
    break

bamfile.close()

In [ ]:
# Install required tools: samtools and bcftools
!apt-get install -y samtools bcftools

# Index the reference genome using samtools
!samtools faidx /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta

# Sort the BAM file if it is not already sorted
!samtools sort /content/sorted_reads.bam -o /content/sorted_reads_sorted.bam

# Generate a pileup file using bcftools mpileup and call variants
!bcftools mpileup -f /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta /content/sorted_reads_sorted.bam -Ou -o /content/raw_variants.bcf

# Call SNPs and indels using bcftools call
!bcftools call -mv -Oz -o /content/variants.vcf /content/raw_variants.bcf

# Filter for SNPs only using bcftools view
!bcftools view -v snps -Oz -o /content/snps_only.vcf /content/variants.vcf

# Display the first few lines of the SNPs-only VCF file
!zcat /content/snps_only.vcf | head

# Optional: Download the filtered SNPs VCF file
from google.colab import files
files.download('/content/snps_only.vcf')

In [ ]:
!pip install vcfpy

In [ ]:
import vcfpy

# Open the VCF file
reader = vcfpy.Reader.from_path('/content/snps_only.vcf')

# Iterate through records and print details of each SNP
for record in reader:
    print(f"Chromosome: {record.CHROM}")
    print(f"Position: {record.POS}")
    print(f"Reference Allele: {record.REF}")
    print(f"Alternate Allele(s): {record.ALT}")
    print(f"Quality: {record.QUAL}")
    print(f"Filter: {record.FILTER}")
    print(f"Info: {record.INFO}")
    print("-" * 50)

    # Break after printing the first few records for demonstration
    break

### What if we do not convert it to BAM?

In [10]:
!head -n 23 /content/aligned_reads.sam

@SQ	SN:AL123456	LN:4411532
@PG	ID:bwa	PN:bwa	VN:0.7.17-r1188	CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/IP/Meeting 4: 23 09 2024/SRR30679840.fastq.gz
SRR30679840.1	0	AL123456	1088325	60	150M	*	0	0	CCTCGGGAAAGCCGGCGCCCAGCAGCCCGGCGGCCCCGGCGAGCCGGTGCAGGCCGCGGGGCAACTCGCCGATCCTTTCCCACTCGTCGACGTGCGGCAGGATCTCGCGCTCGGCAAAGGCGCGCACCGTTTTTCGCAGCTGTTGGCGCT	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:0	MD:Z:150	AS:i:150	XS:i:0
SRR30679840.1	16	AL123456	1088541	60	151M	*	0	0	ACCCAGTCCCTTGGCCTGCGGGTCGAAGCGGGCCTGGTAGGCGACGCCCTGGCCGAGGATTGCCTCGATGACGAAGTTCAGTGCCCGCAGATTCGGCAGCACGTGACGGGTGACGACCAGGCCTGCCGTTTCTGGCAGCAGCTCCTTGAGT	???????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:0	MD:Z:151	AS:i:151	XS:i:0
SRR30679840.2	0	AL123456	3920354	60	

The given line from the SAM file provides detailed alignment information for a single sequencing read (`SRR30683486.1`) against a reference sequence (`AL123456`). Here's a breakdown of each component:

 1\. **Header Lines**

These lines provide meta-information about the reference genome and alignment tools used.

-   **`@SQ SN:AL123456 LN:4411532`**:

    -   **`@SQ`**: Indicates a sequence dictionary entry.
    -   **`SN:AL123456`**: The reference sequence name, in this case, a contig or chromosome labeled `AL123456`.
    -   **`LN:4411532`**: The length of the reference sequence `AL123456` is 4,411,532 base pairs.
-   **`@PG ID:bwa PN:bwa VN:0.7.17-r1188 CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/SRR30683486.fastq.gz`**:

    -   **`@PG`**: Program information.
    -   **`ID:bwa`**: Program ID is `bwa`.
    -   **`PN:bwa`**: Program name is `bwa`.
    -   **`VN:0.7.17-r1188`**: Version of `bwa` used for alignment.
    -   **`CL:bwa mem ...`**: The command line (`CL`) used to run `bwa mem` with the reference file (`mycobacterium_tuberculosis_h37rv_2_contigs.fasta`) and the input read file (`SRR30683486.fastq.gz`).

 2\. **Alignment Line**

The main part of the SAM file, this line provides details about how a single read aligns to the reference genome.

**Line**:

For the first line only:

`SRR30683486.1	0	AL123456	454690	60	150M	*	0	0	GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCGCCGGGCCTGATCGCGCTCTTCTCGTCTCGGGCCGACGACCTTTCGGTCGAGTTGAGCGTGCATCCCACCAGCCGGTTCCGCGAACTGATCTGCTCGCGCGCCGT	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:2	MD:Z:7C51A90	AS:i:140	XS:i:0`

| **Field** | **Description** | **Value** |
| --- | --- | --- |
| **QNAME** | Query name (read identifier). | `SRR30683486.1` |
| **FLAG** | Bitwise flag indicating the read's properties. | `0` (indicates the read is mapped to the forward strand) |
| **RNAME** | Reference sequence name where the read is aligned. | `AL123456` (reference sequence or contig name) |
| **POS** | 1-based leftmost position of alignment. | `454690` (the read starts aligning at position 454,690 on the reference sequence `AL123456`) |
| **MAPQ** | Mapping quality score (0-255). | `60` (high confidence in this alignment) |
| **CIGAR** | Describes the alignment in terms of matches, insertions, deletions, etc. | `150M` (150 bases match the reference exactly, no insertions/deletions) |
| **RNEXT** | Reference name of the mate/next read (if paired). | `*` (no mate information, or single-end read) |
| **PNEXT** | Position of the mate/next read (if paired). | `0` (no mate position, or single-end read) |
| **TLEN** | Observed template length (insert size). | `0` (no template length, or single-end read) |
| **SEQ** | Query sequence (the actual read sequence). | `GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCGCCGGGCCTGATCGCGCTCTTCTCGTCTCGGGCCGACGACCTTTCGGTCGAGTTGAGCGTGCATCCCACCAGCCGGTTCCGCGAACTGATCTGCTCGCGCGCCGT` |
| **QUAL** | Phred-scaled base quality scores. | `??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????` (indicates quality is missing or not available) |
| **Optional Tags** | Additional information about the alignment. | `NM:i:2`, `MD:Z:7C51A90`, `AS:i:140`, `XS:i:0` |

 Detailed Explanation of Key Fields

1.  **QNAME (Query Name)**:

    -   **`SRR30683486.1`**: The unique identifier for the read. This allows you to trace this specific read in your dataset.
2.  **FLAG**:

    -   **`0`**: Indicates the read is mapped to the forward strand with no special properties.
        -   `0` means the read is mapped and is a primary alignment.
3.  **RNAME (Reference Name)**:

    -   **`AL123456`**: The reference sequence or contig to which the read is aligned.
4.  **POS (Position)**:

    -   **`454690`**: The starting position of the read alignment on the reference sequence `AL123456`. This means the first base of the read aligns to position 454,690 of the reference.
5.  **MAPQ (Mapping Quality)**:

    -   **`60`**: Indicates a high confidence in the read alignment. A MAPQ of 60 is the maximum score, signifying the alignment is unique with very high certainty.
6.  **CIGAR (Compact Idiosyncratic Gapped Alignment Report)**:

    -   **`150M`**: This indicates that all 150 bases in the read match exactly to the reference genome without any insertions, deletions, or clipping.
        -   `150M` means 150 bases are aligned with either matches or mismatches but no gaps or clipping.
7.  **SEQ (Sequence)**:

    -   The sequence of the read that has been aligned to the reference.
    -   **`GTTGCTGAGCATCGCC...`**: This is the nucleotide sequence of the read.
8.  **QUAL (Quality Scores)**:

    -   The quality scores for each base in the read. Typically, these are ASCII-encoded Phred scores representing the confidence in each base call.
    -   **`????????????...`**: In this case, it appears as `?`, which may indicate missing or placeholder quality scores, suggesting the quality information is not present.
9.  **Optional Tags**:

    -   Provide additional details about the alignment:

    -   **`NM:i:2`**: Number of mismatches between the read and the reference sequence. Here, there are 2 mismatches.

    -   **`MD:Z:7C51A90`**: Mismatch descriptor tag:

        -   `7C51A90` means:
            -   `7`: First 7 bases match exactly.
            -   `C`: Mismatch at the 8th position (reference has `C`).
            -   `51`: Next 51 bases match exactly.
            -   `A`: Mismatch at the 60th position (reference has `A`).
            -   `90`: Last 90 bases match exactly.
    -   **`AS:i:140`**: Alignment score, a measure of how well the read aligns to the reference. Higher scores indicate better alignments.

    -   **`XS:i:0`**: Suboptimal alignment score. If there were multiple potential alignments for this read, this would show the score of the next best alignment. Here, `0` suggests no suboptimal alignment.

 Summary

This alignment line provides a detailed view of how the read `SRR30683486.1` aligns to the reference sequence `AL123456`:

-   The read aligns starting at position `454690` on the reference with a high-quality score of `60`.
-   The CIGAR string `150M` indicates that all 150 bases of the read match exactly to the reference sequence.
-   There are 2 mismatches between the read and the reference, as indicated by the `NM:i:2` tag.
-   The `MD:Z:7C51A90` tag specifies the exact positions of these mismatches.

This information is crucial for understanding the accuracy and quality of the alignment, which is necessary for downstream analyses like variant calling or gene expression studies. If you need more details or specific interpretations, feel free to ask!

In [11]:
!cat /content/aligned_reads.sam

Streaming output truncated to the last 5000 lines.
SRR30679840.730920	0	AL123456	679820	60	144M	*	0	0	ATCGAGGACAACACCACAATCACCGGCTCGGCGACCAGCGGCGGGTTCTGCGGCAAGAAGAACCTGTACACGGTGTACTTCGCCATGAAGTTCAGCCAGCCGTTCAGCTCGTATGGCACCTGGGACGGCTACGCGGTCTATCCC	????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:0	MD:Z:144	AS:i:144	XS:i:21
SRR30679840.730921	0	AL123456	348428	60	150M	*	0	0	CTCTATGGCTTCGGCCGCGATAACACCGATCAGTTCAACACCCGTAATCCGTTGCTGGCCAGGGTAACCGAACTGGTCAACGTGGGCCTTGCCTACGGGATCCACGTGATCATTACCACGCCGAGCTGGCTGGAAGTGCCGTTGGCGATG	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:0	MD:Z:150	AS:i:150	XS:i:0
SRR30679840.730921	16	AL123456	348734	60	150M	*	0	0	GCTCCAGAACTGGACGCGCAAACAAACCCGGTGGCCGCGATCAACGCCCGCTACCCCGGCATGGCGGCTCCCCCGGTTCGGTTGTTGCCCACCAACCTTGCGCCGCACGCCGTCGGCGAACTGTATCGGGGTCCCGAC

Total number of lines

In [12]:
!wc -l /content/aligned_reads.sam

1474272 /content/aligned_reads.sam


In [ ]:
574362*150/(2*4.4*10**6)

9.790261363636363

To find the total number of discrepancies (mismatches, insertions, and deletions) between the reads in your SAM file and the reference genome, you can use the samtools tool along with some additional processing.

In [ ]:
!apt-get install -y samtools
!samtools view -Sb /content/aligned_reads.sam > /content/aligned_reads.bam
#!samtools view -Sb /content/aligned_reads.sam > /content/aligned_reads.bam
!samtools view /content/aligned_reads.bam | awk '{for(i=12;i<=NF;i++) if($i ~ /^NM:i:/) {split($i,a,":"); total += a[3]}} END {print "Total number of discrepancies:", total}'

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
samtools is already the newest version (1.13-4).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.
Total number of discrepancies: 249797


In [ ]:
import pandas as pd

# Read the first 20 lines of the SAM file, skipping lines with parsing errors
sam_file = '/content/aligned_reads.sam'
sam_data = pd.read_csv(
    sam_file,
    sep='\t',
    comment='@',
    header=None,
    names=None,
    on_bad_lines='skip',  # Skip lines with mismatched columns  # Read the first 20 lines
)

# Display the DataFrame
sam_data

0   1         2        3   4        5  6   7   8   \
0            SRR30683486.1   0  AL123456   454690  60     150M  *   0   0   
1            SRR30683486.1  16  AL123456   455061  60     150M  *   0   0   
2            SRR30683486.2  16  AL123456  1425788  60     151M  *   0   0   
3            SRR30683486.2   0  AL123456  1425435  60     148M  *   0   0   
4            SRR30683486.3  16  AL123456  3102028  60     151M  *   0   0   
...                    ...  ..       ...      ...  ..      ... ..  ..  ..   
565483  SRR30683486.286531  16  AL123456  4173889  60     146M  *   0   0   
565484  SRR30683486.286532  16  AL123456  2798987  60     150M  *   0   0   
565485  SRR30683486.286532   0  AL123456  2798544  60     114M  *   0   0   
565486  SRR30683486.286533  16  AL123456  3633106  60     151M  *   0   0   
565487  SRR30683486.286533   0  AL123456  3632971  60  116M35S  *   0   0   

                                                       9   \
0       GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCG...   
1       CGATTCCGGAGTCCCAGCAACGGATCTTCCAGAGCGATGCCGCCGC...   
2       CATCGACGGCCGGCAGGTTGGCGCGCAGCCAGCCACCGGCCAGCCC...   
3       CCTCTACCGTGGCACGTGAAAGCCGACCAGCGCGGCACACCCCGGT...   
4       ATCGGCCGGCTGGTCGTCCACGAGAATCAGCCGCCCGGCCTTCCGC...   
...                                                   ...   
565483  ATCTGACGCTGCGAGAGCTGGCCAAATCAGCTCATGTCAAGCGATA...   
565484  CACGCCACAACGGCTGCAGACATGATGCCAACAGCAACGCGTGGCC...   
565485  AAGGAAGGCATCACCAATCTCGTCGGCGAGGCTTCGTGCCTCCTCG...   
565486  AACGTGACGATCACGAGCGCACCCCAGTCGAGCTGCAGCGATCCGA...   
565487  GGCGAGTAGTTGGCGGCTCTGATGTAGAAGGCACCGACGACCACCA...   

                                                       10      11  \
0       ??????????????????????????????????????????????...  NM:i:2   
1       ??????????????????????????????????????????????...  NM:i:1   
2       ??????????????????????????????????????????????...  NM:i:0   
3       ??????????????????????????????????????????????...  NM:i:0   
4       ??????????????????????????????????????????????...  NM:i:0   
...                                                   ...     ...   
565483  ??????????????????????????????????????????????...  NM:i:0   
565484  ??????????????????????????????????????????????...  NM:i:0   
565485  ??????????????????????????????????????????????...  NM:i:1   
565486  ??????????????????????????????????????????????...  NM:i:1   
565487  ??????????????????????????????????????????????...  NM:i:3   

                     12        13       14  
0          MD:Z:7C51A90  AS:i:140   XS:i:0  
1            MD:Z:75C74  AS:i:145   XS:i:0  
2              MD:Z:151  AS:i:151   XS:i:0  
3              MD:Z:148  AS:i:148   XS:i:0  
4              MD:Z:151  AS:i:151   XS:i:0  
...                 ...       ...      ...  
565483         MD:Z:146  AS:i:146  XS:i:20  
565484         MD:Z:150  AS:i:150   XS:i:0  
565485       MD:Z:6T107  AS:i:109   XS:i:0  
565486      MD:Z:119A31  AS:i:146   XS:i:0  
565487  MD:Z:17C78G6A12  AS:i:107   XS:i:0  

[565488 rows x 15 columns]

In [ ]:
565488-286533*2

-7578

In [1]:
sum = 0
for i in sam_data[11]:
  sum+= int(i[-1])
print(sum) #== #249797

NameError: name 'sam_data' is not defined

# Reading and using Picard

### 1\. **Sorting and Merging BAM/SAM Files**

-   **Sorting**: Think of sorting as arranging your data in a particular order. For BAM/SAM files, this usually means ordering the reads based on their position in the genome. This makes it easier to find and analyze specific regions in your data.
-   **Merging**: If you have multiple files containing parts of your data, merging them combines all these parts into a single, complete file. This is useful when you have data from different experiments or runs and want to analyze them together.

### 2\. **Marking Duplicates**

-   During sequencing, sometimes the same piece of DNA gets copied multiple times. Picard can find these duplicate copies and mark them. By marking duplicates, you can ignore them in your analysis, so your results aren't skewed by counting the same piece of DNA more than once.

### 3\. **Collecting Quality Metrics**

-   Picard can check the quality of your sequencing data by calculating various statistics. For example, it can tell you the average size of DNA fragments, how well your reads align to the reference genome, and if there are any unusual patterns. This helps ensure your data is good enough for further analysis.

### 4\. **File Conversion**

-   Picard can change the format of your data files. For example, it can switch between BAM (a compressed, efficient format) and SAM (a plain-text version) formats. This is helpful when different tools require different formats.

### 5\. **Manipulating Headers**

-   The header of a BAM/SAM file contains important information about the data, like the sample name and sequencing details. Picard allows you to change this information, such as adding new details or correcting mistakes, so that your files have accurate and complete metadata.

Picard is a Java-based tool, so you need to install Java first.

In [ ]:
!apt-get install -y openjdk-11-jdk

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java libatk-wrapper-java-jni libfontenc1
  libice-dev libsm-dev libxkbfile1 libxt-dev libxtst6 libxxf86dga1 openjdk-11-jre x11-utils
Suggested packages:
  libice-doc libsm-doc libxt-doc openjdk-11-demo openjdk-11-source visualvm mesa-utils
The following NEW packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java libatk-wrapper-java-jni libfontenc1
  libice-dev libsm-dev libxkbfile1 libxt-dev libxtst6 libxxf86dga1 openjdk-11-jdk openjdk-11-jre
  x11-utils
0 upgraded, 14 newly installed, 0 to remove and 49 not upgraded.
Need to get 5,517 kB of archives.
After this operation, 15.8 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-core all 2.37-2build1 [1,041 kB]
Get:2 http://archive.ubuntu.com/ubun

You can download the Picard JAR file from the official Broad Institute website.

In [ ]:
!wget -O picard.jar https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar

--2024-09-23 05:14:02--  https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar
Resolving github.com (github.com)... 140.82.116.3
Connecting to github.com (github.com)|140.82.116.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240923%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240923T051402Z&X-Amz-Expires=300&X-Amz-Signature=6646d88162e2712a44bf98f28656cdb64bf9b900d1e7c50ef41e8f41b9ad6cc9&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dpicard.jar&response-content-type=application%2Foctet-stream [following]
--2024-09-23 05:14:02--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=r

In [ ]:
!java -jar picard.jar SortSam \
    INPUT=/content/aligned_reads.bam \
    OUTPUT=/content/sorted_reads.bam \
    SORT_ORDER=coordinate

INFO	2024-09-23 05:25:41	SortSam	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    SortSam -INPUT /content/aligned_reads.bam -OUTPUT /content/sorted_reads.bam -SORT_ORDER coordinate
**********


05:25:43.016 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Mon Sep 23 05:25:43 UTC 2024] SortSam INPUT=/content/aligned_reads.bam OUTPUT=/content/sorted_reads.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false
[Mon Sep 23 05

Key Points in the Log:

1.  **Command Line Syntax Change Notification**:

    -   Picard is notifying you that the command line syntax is changing. They are moving to a new syntax format. The example given in the log shows how the new command syntax would look:

        bash

        `SortSam -INPUT /content/aligned_reads.bam -OUTPUT /content/sorted_reads.bam -SORT_ORDER coordinate`

2.  **Library Loading**:

    -   `INFO NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!...`
    -   This indicates that Picard is loading native libraries required for compression and decompression operations, which are used to handle the BAM file format efficiently.
3.  **Execution Information**:

    -   The line `[Mon Sep 23 05:25:43 UTC 2024] SortSam INPUT=/content/aligned_reads.bam OUTPUT=/content/sorted_reads.bam ...`
    -   This shows the command being executed with various parameters:
        -   `INPUT`: The input BAM file.
        -   `OUTPUT`: The output file to which the sorted reads are written.
        -   `SORT_ORDER`: The order in which the reads are sorted (in this case, `coordinate`).
        -   `VERBOSITY=INFO`: The level of logging information.
        -   `VALIDATION_STRINGENCY=STRICT`: How strictly Picard checks the input file for errors.
        -   Other parameters control compression, indexing, memory usage, and more.
4.  **Non-Increasing Record Positions**:

    -   `INFO 2024-09-23 05:25:43 SortSam Seen many non-increasing record positions. Printing Read-names as well.`
    -   This message suggests that the input BAM file had records (reads) that were not in the expected order (which is why you're sorting them). This is typical when sorting BAM files.
5.  **Processing Complete**:

    -   `INFO 2024-09-23 05:25:51 SortSam Finished reading inputs, merging and writing to output now.`
    -   This indicates that Picard has finished reading the input BAM file, merged the reads in the correct order, and is now writing the sorted reads to the output BAM file.
6.  **Execution Finished**:

    -   `[Mon Sep 23 05:25:56 UTC 2024] picard.sam.SortSam done. Elapsed time: 0.22 minutes.`
    -   This means the entire sorting process is complete, and it took around 0.22 minutes (about 13 seconds) to finish the task.
7.  **Memory Usage**:

    -   `Runtime.totalMemory()=124780544`
    -   This is the amount of memory (in bytes) allocated to the Java Virtual Machine (JVM) during the execution.

### Summary:

The log indicates that the `SortSam` command successfully sorted the input BAM file, despite encountering many non-increasing record positions (which is normal for unsorted BAM files). The new syntax notification is just a heads-up for future command usage. The operation was completed in around 13 seconds, and everything seems to have worked as expected.

In [ ]:
!java -jar picard.jar MarkDuplicates \
    INPUT=/content/sorted_reads.bam \
    OUTPUT=/content/dedup_reads.bam \
    METRICS_FILE=/content/duplication_metrics.txt

INFO	2024-09-23 05:48:54	MarkDuplicates	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    MarkDuplicates -INPUT /content/sorted_reads.bam -OUTPUT /content/dedup_reads.bam -METRICS_FILE /content/duplication_metrics.txt
**********


05:48:55.405 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Mon Sep 23 05:48:55 UTC 2024] MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATE

The log you've shared is the detailed output of running the `MarkDuplicates` command using Picard. Here's a step-by-step explanation of the information provided in the log:

Key Points in the Log:

1.  **Command Line Syntax Change Notification**:

    -   Similar to `SortSam`, Picard is informing you that the command line syntax is changing. The log provides the new format of the `MarkDuplicates` command:

        bash

        Copy code

        `MarkDuplicates -INPUT /content/sorted_reads.bam -OUTPUT /content/dedup_reads.bam -METRICS_FILE /content/duplication_metrics.txt`

2.  **Library Loading**:

    -   `INFO NativeLibraryLoader - Loading libgkl_compression.so...`
    -   Picard is loading native libraries for compression, which are required to handle BAM files efficiently.
3.  **Command Execution Details**:

    -   The command being executed is:

        javascript

        Copy code

        `MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt ...`

    -   Key parameters used:
        -   `INPUT`: Specifies the input sorted BAM file.
        -   `OUTPUT`: Specifies the output BAM file with duplicates marked.
        -   `METRICS_FILE`: Specifies the file where duplication metrics (e.g., number of duplicates) are written.
        -   Various other parameters control aspects like memory usage, duplicate scoring, and how read names are parsed.
4.  **Start of Processing**:

    -   `INFO MarkDuplicates Start of doWork freeMemory: 24723776; totalMemory: 31457280; maxMemory: 3403677696`
    -   Shows initial memory allocation for the Java process before starting the duplicate marking operation.
5.  **Reading and Constructing Read End Information**:

    -   `INFO MarkDuplicates Reading input file and constructing read end information.`
    -   Picard is reading the input BAM file to gather information about the reads' ends (the start and end positions of each read) to identify duplicates.
6.  **Non-Integer Read Name Warning**:

    -   `WARNING AbstractOpticalDuplicateFinderCommandLineProgram ... Read name: SRR30683486.118524. Cause: String 'SRR30683486.118524' did not start with a parsable number.`
    -   This warning indicates that a part of a read name, which Picard expected to contain an integer, did not. This is usually not a critical issue but indicates that the read naming convention is not as expected by Picard.
7.  **Number of Records Read**:

    -   `INFO MarkDuplicates Read 568784 records. 0 pairs never matched.`
    -   Picard read 568,784 records from the BAM file and did not encounter any unmatched read pairs, which means all paired-end reads were properly aligned.
8.  **Memory Usage Information**:

    -   Several memory usage statistics are logged throughout the process, indicating how much memory Picard is using as it processes the file.
9.  **Duplicate Detection**:

    -   `INFO MarkDuplicates Traversing read pair information and detecting duplicates.`
    -   Picard is now checking each read pair for duplicates.
10. **Duplicates Marked**:

    -   `INFO MarkDuplicates Marking 102504 records as duplicates.`
    -   102,504 records were identified and marked as duplicates.
11. **Optical Duplicate Detection**:

    -   `INFO MarkDuplicates Found 0 optical duplicate clusters.`
    -   No optical duplicates (duplicates caused by the imaging system of the sequencer) were found. This is good, as it suggests no imaging artifacts.
12. **Writing Complete**:

    -   `INFO MarkDuplicates Writing complete. Closing input iterator.`
    -   Picard has finished writing the output BAM file with duplicates marked and is closing the input and output files.
13. **Duplicate Index Cleanup**:

    -   `INFO MarkDuplicates Duplicate Index cleanup.`
    -   Picard is cleaning up any temporary indices it created during the processing.
14. **Final Memory Statistics and Output Closure**:

    -   Shows the memory usage statistics before and after closing the output file.
15. **Execution Complete**:

    -   `[Mon Sep 23 05:49:11 UTC 2024] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.28 minutes.`
    -   The `MarkDuplicates` command has finished execution in approximately 0.28 minutes (about 17 seconds).
16. **Final Memory Usage**:

    -   `Runtime.totalMemory()=28311552`
    -   This shows the final memory allocated to the JVM after completing the command.

Summary:

The `MarkDuplicates` command successfully identified and marked duplicates in the BAM file. It read 568,784 records and marked 102,504 of them as duplicates. The process took around 17 seconds to complete, and no optical duplicates were found. The log also includes detailed information on memory usage and processing steps, indicating that everything completed successfully without critical errors.


In [ ]:
!java -jar picard.jar SamFormatConverter \
    INPUT=/content/aligned_reads.bam \
    OUTPUT=/content/aligned_reads.sam

INFO	2024-09-23 05:47:08	SamFormatConverter	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    SamFormatConverter -INPUT /content/aligned_reads.bam -OUTPUT /content/aligned_reads.sam
**********


05:47:09.812 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Mon Sep 23 05:47:09 UTC 2024] SamFormatConverter INPUT=/content/aligned_reads.bam OUTPUT=/content/aligned_reads.sam    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false
[Mon Sep 23 05:47:09 UTC

 Command Breakdown:

`!java -jar picard.jar AddOrReplaceReadGroups\
    INPUT=/content/aligned_reads.bam\
    OUTPUT=/content/rg_aligned_reads.bam\
    RGID=1\
    RGLB=lib1\
    RGPL=illumina\
    RGPU=unit1\
    RGSM=sample`

-   `!java -jar picard.jar AddOrReplaceReadGroups \`:

    -   This is the main command to run Picard's `AddOrReplaceReadGroups` tool using the Java command-line interface.
-   `INPUT=/content/aligned_reads.bam \`:

    -   Specifies the path to the input BAM file (`/content/aligned_reads.bam`) where the read group information will be added or replaced.
-   `OUTPUT=/content/rg_aligned_reads.bam \`:

    -   Specifies the path to the output BAM file (`/content/rg_aligned_reads.bam`) with the updated read group information.
-   `RGID=1 \`:

    -   `RGID` stands for "Read Group ID." It uniquely identifies a read group within the BAM file. Here, it is set to `1`.
-   `RGLB=lib1 \`:

    -   `RGLB` stands for "Read Group Library." It specifies the library preparation kit or process used. Here, it is set to `lib1`.
-   `RGPL=illumina \`:

    -   `RGPL` stands for "Read Group Platform." It specifies the sequencing platform used. In this case, `illumina` is specified, which is a commonly used sequencing technology.
-   `RGPU=unit1 \`:

    -   `RGPU` stands for "Read Group Platform Unit." It typically refers to the run or flow cell barcode used for sequencing. Here, it is set to `unit1`.
-   `RGSM=sample`:

    -   `RGSM` stands for "Read Group Sample." It specifies the name of the sample being sequenced. Here, it is set to `sample`.

 Why Is This Important?

1.  **Metadata Annotation**: Adding read group information helps in keeping track of different sequencing runs, libraries, and samples, which is crucial for organizing and analyzing complex datasets.

2.  **Downstream Analysis**: Many tools, like GATK (Genome Analysis Toolkit) for variant calling, require read group information to differentiate between data from different samples or sequencing runs within a single BAM file.

3.  **Quality Control**: By including details about the platform and library, researchers can perform quality checks and identify issues specific to a particular run or library preparation.

 Summary:

This command adds or updates metadata in your BAM file to include specific details about the sequencing process and sample. This information is essential for proper data management and accurate downstream analysis.